crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1718,7 +1736,9 @@ impl<'db> TypeInferenceBuilder<'db> {
 
             // Inference of bases deferred in stubs
             // TODO also defer stringified generic type parameters
-            if self.are_all_types_deferred() {
+            if self.are_all_types_deferred()
+                || class_node.bases().iter().any(contains_string_literal)
```

Comment:
Can be a TODO for now, but I suspect that deferring _all_ inference of class bases if any of them contain any string literal is too big a hammer here, and we will need to do something more complex. In a case like this:

```py
class B[T]: pass

class C(B["Forward"]): pass

class Forward: pass

B = "oops"
```

We have to infer the type of `B` at the point of use (and not pick up the later binding of `B`), but we have to use a deferred lookup for `"Forward"`. This will be a bit awkward, as it means we have to keep track of unspecialized `B` somewhere, until we actually resolve `"Forward"` and can specialize.